File: ssim/Math/KochSnowflake/kochsnow.py
import numpy as np
import matplotlib.pyplot as plt

def koch_snowflake(iterations=4,anti=False,display_iter=False,display=True,vectorize=True):
    ''' Compute the points that make the koch snowflake
    Parameters
    ----------
    iterations : int,
        Number of iterations for the koch snowflake
    anti : boolean,
        if false compute the koch snowflake, if true compute the antisnowflake. Defaults to False.
    display : boolean,
        Wether to display the koch snowflake
    display_iter : boolean,
        Wether to display the koch snowflake at each iteration
    vectorize : boolean,
        If true use a vectorize version of the algo (much faster), Defaults to True.

    Yields
    ------
    points: array,
        list of points that makes the koch snowflake
    '''
    # If anti snowflake, angle is -pi/3
    if anti :
        angle_mov = -np.pi/3
    else :
        angle_mov = np.pi/3

    # Starting points
    tan_angle = np.tan(np.pi/3)
    points = [[0,0],[3**iterations/2,tan_angle*3**iterations/2],[3**iterations,0],[0,0]]

    def draw(points):
        points_np = np.array(points)
        plt.figure(figsize=(10,10))
 
        plt.plot(points_np[:,0],points_np[:,1])
        plt.axis('square')
    
    if vectorize : 
        for _ in range(iterations-1):
            new_points = [] 
            
            points_next = np.roll(points,-1,axis=0)[:-1]
            diff = points_next - points[:-1]

            side_eq = np.sqrt(np.sum(diff**2,axis=1))/3  # Signed len of new triangle side (/3)
            angle = np.arctan2(diff[:,1],diff[:,0])

            point1 = points[:-1] + diff/3 # First point of new triangle (first third of line)

            point2 = np.zeros(point1.shape)
            point2[:,0] = point1[:,0] + side_eq * np.cos(angle+angle_mov)
            point2[:,1] = point1[:,1] + side_eq * np.sin(angle+angle_mov)
            
            point3 = points[:-1] + 2*diff/3 # First point of new triangle (first third of line)

            # Concatenate the arrays
            new_points = np.zeros((point1.shape[0]*4+1,2))
            idx = np.arange(new_points.shape[0])
            idx_tmp = np.where(idx%4==0)[0]
            new_points[idx_tmp] = points
            idx_tmp = idx_tmp[:-1]
            idx_tmp +=1
            new_points[idx_tmp] = point1
            idx_tmp +=1
            new_points[idx_tmp] = point2
            idx_tmp +=1
            new_points[idx_tmp] = point3

            points  = new_points

            if display_iter : 
                draw(points)

    else : ## old non vectorize version (much slower)

        for _ in range(iterations-1):
            new_points = [] 
            for i in range(len(points)-1) : 
                new_points.append(points[i]) # Add old point to new list

                x_diff = points[i+1][0]-points[i][0]
                y_diff = points[i+1][1]-points[i][1]

                side_eq = np.sqrt( (x_diff)**2 + (y_diff)**2 )/3  # Signed len of new triangle side (/3)
                angle = np.arctan2(y_diff,x_diff) # angle between 2 points
                
                # First point of new triangle (first third of line)
                new_x = points[i][0] + x_diff/3
                new_y = points[i][1] + y_diff/3
                new_points.append([new_x,new_y])

                # Top of new triangle
                new_x += side_eq * np.cos(angle+angle_mov)
                new_y += side_eq * np.sin(angle+angle_mov)
                new_points.append([new_x,new_y])

                # Third and last point of new triangle (first third of line)
                new_x = points[i][0] + 2*x_diff/3
                new_y = points[i][1] + 2*y_diff/3
                new_points.append([new_x,new_y])
                
            new_points.append(points[-1]) # Add last point
            points = new_points

            if display_iter : 
                draw(points)

    if display and not display_iter:
        draw(points)

    if display or display_iter : 
        plt.show()
    return points

def main():
    koch_snowflake(6)

if __name__ == "__main__":
    main()

File: ssim/Math/KochSnowflake/kochsnow.cpp
#define _USE_MATH_DEFINES
#include<iostream>
#include<cmath>
// #include<matplot/matplot.h>

#include<fstream>

using namespace std;
// using namespace matplot;

void koch_snowflake(bool anti,int iter){

    // Variable definition

    float angle_mov;
    float *xpoints;
    float *ypoints;
    int len_points = 4;
    int len_new_points;
    float xdiff;
    float ydiff;
    float side_eq;
    float angle;


    if (anti == false)
    {
        angle_mov = - M_PI /3;
    }
    else
    {
        angle_mov = M_PI /3;
    }

    // define first points
    float tan_angle = tan(M_PI/3);
    cout << tan_angle << "\n";
    float xinit[4] = {0.0f, pow(3.0f,(float) iter/2), pow(3.0f,(float) iter), 0.0f};
    xpoints = xinit;
    float yinit[4] = {0.0f, tan_angle * pow(3.0f,(float) iter/2), 0.0f , 0.0f};
    ypoints = yinit;

    cout<< "\nONE : " ;
        
        for (int k = 0; k <= len_points; k++)
        {
            cout << "(" <<xpoints[k] << ", "<< ypoints[k] << "),";
        }
        cout << "\n";


    // TODO draw void function

    for (int i = 0; i < iter; i++)
    {      
        len_new_points = len_points +  (len_points-1)*3;

        float xnew_points[len_new_points];
        float ynew_points[len_new_points];

        xnew_points[0] = xpoints[0];
        ynew_points[0] = ypoints[0];

        for (int j = 0; j < len_points; j++)
        {
            
            // First point of the new triangle
            xdiff = xpoints[j+1] - xpoints[j];
            ydiff = ypoints[j+1] - ypoints[j];

            side_eq = sqrt( pow(xdiff,2) + pow(ydiff,2) )/3;
            angle = atan2(ydiff,xdiff);

            xnew_points[j*4 + 1] = xpoints[j] + xdiff/3;
            ynew_points[j*4 +1] = ypoints[j] + ydiff/3;

            // Top of the triangle
            xnew_points[j*4+2] = xnew_points[j*4 + 1] + side_eq * cos(angle+angle_mov);
            ynew_points[j*4+2] = ynew_points[j*4 + 1] + side_eq * sin(angle+angle_mov);

            // Third point of the triangle
            xnew_points[j*4+3] = xpoints[j] + 2 * xdiff/3;
            ynew_points[j*4+3] = ypoints[j] + 2 * ydiff/3;

            // Old point
            xnew_points[j*4 +4] = xpoints[j];
            ynew_points[j*4 +4] = ypoints[j];

        }
    
        len_points = len_new_points;

        cout<< "\n\n UPDATE : " ;
        
        for (int k = 0; k <= len_new_points; k++)
        {
            cout << "(" <<xnew_points[k] << ", "<< ynew_points[k] << "),";
        }
        cout << "\n";
         
        
        xpoints = xnew_points;
        ypoints = ynew_points;
    }    
    
}

int main() 
{      
    int depth;
    bool anti;

    cout << "Depth ? "; 
    cin >> depth;

    koch_snowflake(false,depth);  
      
    return 0; 
} 